File: vscode/src/core/localizations/langFileScanner.ts
import * as fs from 'fs';
import * as path from 'path';

/** Port of `dev.nylo.plugin.localizations.scan.LangFileScanner`. */

export const LANG_DIR = 'lang';

export interface LangFile {
  /** Locale code, e.g. `en`. */
  code: string;
  filePath: string;
}

/** Discovers `lang/<code>.json` files under [projectDir], sorted by code. */
export function scanLangFiles(projectDir: string): LangFile[] {
  const langDir = path.join(projectDir, LANG_DIR);
  let names: string[];
  try {
    if (!fs.statSync(langDir).isDirectory()) {
      return [];
    }
    names = fs.readdirSync(langDir);
  } catch {
    return [];
  }

  const files: LangFile[] = [];
  for (const name of names) {
    if (!name.endsWith('.json')) {
      continue;
    }
    const full = path.join(langDir, name);
    try {
      if (!fs.statSync(full).isFile()) {
        continue;
      }
    } catch {
      continue;
    }
    files.push({ code: name.slice(0, -'.json'.length), filePath: full });
  }
  files.sort((a, b) => (a.code < b.code ? -1 : a.code > b.code ? 1 : 0));
  return files;
}

File: intellij/src/main/kotlin/dev/nylo/plugin/localizations/scan/LangFileScanner.kt
package dev.nylo.plugin.localizations.scan

import com.intellij.openapi.project.Project
import java.io.File

/**
 * Discovers Nylo locale files under the project's `lang/` folder — one `<code>.json` per language.
 *
 * Mirrors [dev.nylo.plugin.screenshots.project.LocaleScanner] but returns the backing files (not just
 * codes), since the localizations feature needs to read their contents. The `File` overload keeps it
 * unit-testable without a `Project`.
 */
object LangFileScanner {
    fun scan(project: Project): List<LangFile> {
        val basePath = project.basePath ?: return emptyList()
        return scan(File(basePath))
    }

    fun scan(projectDir: File): List<LangFile> {
        val langDir = File(projectDir, "lang")
        if (!langDir.isDirectory) return emptyList()
        val files = langDir.listFiles { f -> f.isFile && f.extension == "json" } ?: return emptyList()
        return files.map { LangFile(it.nameWithoutExtension, it) }.sortedBy { it.code }
    }

    /** The project's `lang/` directory (whether or not it exists). */
    fun langDir(project: Project): File? = project.basePath?.let { File(File(it), "lang") }
}
